### DISCUSSION TITLE
How to get a valid molecule from findMCS result.

### DISCUSSION THREAD

**Discussion by kexul** [rdkit/rdkit#4019] [Q&A] [answered]
Hi, I'm using the following code to find the mcs of two molecule, and then form a new molecule from the mcs result. 
```
p1 = Chem.MolFromSmiles('NCc1cccc(C2CCN(C(=O)c3cn(-c4cccnc4)c4ccccc34)CC2)c1')
p2 = Chem.MolFromSmiles('NCc1cccc(C2CCN(C(=O)c3c[nH]c4ccccc34)CC2)c1')

res = rdFMCS.FindMCS([p2, p1], threshold=0.9, completeRingsOnly=True, timeout=300)
patt = Chem.MolFromSmarts(res.smartsString)
common_core = AllChem.ReplaceSidechains(p1, patt)
common_core = AllChem.DeleteSubstructs(common_core, Chem.MolFromSmiles('*'))
common_core.UpdatePropertyCache()

common_smi = Chem.MolToSmiles(common_core)

new_mol = Chem.MolFromSmiles(common_smi)
```

But in the last line of the code, I can't get a valid SMILES and the program will complain: 
```
[16:29:26] Can't kekulize mol.  Unkekulized atoms: 13 14 15 16 17 18 19 20 21
```

Is there some way to get a valid molecule? Thanks!

**Comment by ptosco** [comment, marked as answer]
@kexul `DeleteSubstructs` will chop off the dummy atoms but will not adjust the implicit H count on the indole nitrogen, hence the kekulization failure. A better strategy is replacing the dummies with hydrogens, and then calling `RemoveAllHs` which will remove the explicit  hydrogens and adjust the implicit H count; this will give you a valid molecule:
```
patt = Chem.MolFromSmarts(res.smartsString)
common_core = Chem.ReplaceSidechains(p1, patt)
for a in common_core.GetAtoms():
    if (a.GetAtomicNum() == 0):
        a.SetAtomicNum(1)
common_core = Chem.RemoveAllHs(common_core)
common_core
```
![image](https://user-images.githubusercontent.com/5244385/114265417-f6fe6500-99f0-11eb-9e87-3c28cca625b4.png)

**Reply by kexul** [reply]
Thanks so much for your detailed answer!

**Reply by kexul** [reply]
Hi @ptosco , I found that my rdkit did not have the `RemoveAllHs()` function. When I use `RemoveHs()` instead, it complained: 
```
rdkit.Chem.rdchem.MolSanitizeException: non-ring atom 16 marked aromati
```
From some dependency reason, I'm not able to upgrade my rdkit version, Is it possible to do it with rdkit version `2019.09.3`?

**Reply by ptosco** [reply]
It looks like `ReplaceSidechains` does not clear the aromatic flag from the dummy atom; you can do it manually to solve the problem with `RemoveHs`:
```
for a in common_core.GetAtoms():
    if (a.GetAtomicNum() == 0):
        a.SetAtomicNum(1)
        a.SetIsotope(0)
        a.SetIsAromatic(False)
    print (a.GetIdx(), a.GetSymbol())
common_core = Chem.RemoveHs(common_core)
```

**Reply by kexul** [reply]
Thanks for your help! @ptosco  It helps me a lot!

**Reply by kexul** [reply]
Hi, @ptosco , I got another bad case, which raised the error:
```
rdkit.Chem.rdchem.MolSanitizeException: non-ring atom 8 marked aromatic
```
, here is my code: 
```
p1 = Chem.MolFromSmiles('Cc1ccc2c(c1)c(-c1cncc(N)c1)c1n2C[C@@](NC(=O)c2c(Cl)cc(-n3cnnc3)cc2Cl)(c2ccccc2)CC1')
p2 = Chem.MolFromSmiles('Cc1ccc2c(c1)c(C)c1n2C[C@@](NC(=O)c2c(Cl)cc(-n3cnnc3)cc2Cl)(c2ccccc2)CC1')

res = rdFMCS.FindMCS([p2, p1], completeRingsOnly=True, timeout=300)
patt = Chem.MolFromSmarts(res.smartsString)
common_core = AllChem.ReplaceSidechains(p1, patt)
print(Chem.MolToSmiles(common_core))
for a in common_core.GetAtoms():
    if a.GetAtomicNum() == 0:
        a.SetAtomicNum(1)
        a.SetIsotope(0)
        a.SetIsAromatic(False)
        print(a.GetIdx())
common_core = Chem.RemoveHs(common_core)
```

**Reply by ptosco** [reply]
@kexul There are further issues in `ReplaceSidechains` that I will address later in a PR.
As a quick workaround to fix your current case:
```
for a in common_core.GetAtoms():
    if (not a.IsInRing()):
        a.SetIsAromatic(False)
    if a.GetAtomicNum() == 0:
        a.SetAtomicNum(1)
        a.SetIsotope(0)
        a.SetIsAromatic(False)
        print(a.GetIdx())
```

**Reply by kexul** [reply]
Hi @ptosco , thanks for your workaround, It fix that case. Unfortunately, I got another bad case: 
```
Cc1ccc2c(c1)c(/C=C/C(=O)O)c1n2C[C@@](NC(=O)c2c(Cl)cc(-n3cnnc3)cc2Cl)(c2ccccc2)CC1
Cc1ccc2c(c1)c(C)c1n2C[C@@](NC(=O)c2c(Cl)cc(-n3cnnc3)cc2Cl)(c2ccccc2)CC1
```
which raised an error: 
```
rdkit.Chem.rdchem.MolSanitizeException: Explicit valence for atom # 9 H, 2, is greater than permitted
```

**Reply by ptosco** [reply]
@kexul Yes, I found and fixed that one yesterday :-)
Current workaround:
```
for b in common_core.GetBonds():
    if (b.GetBeginAtom().GetAtomicNum() == 0 or b.GetEndAtom().GetAtomicNum() == 0):
        b.SetBondType(Chem.BondType.SINGLE)
        b.SetIsAromatic(False)
for a in common_core.GetAtoms():
    if (not a.IsInRing()):
        a.SetIsAromatic(False)
    if a.GetAtomicNum() == 0:
        a.SetAtomicNum(1)
        a.SetIsotope(0)
        a.SetIsAromatic(False)
        print(a.GetIdx())
```
Hopefully this will allow you to complete your task without further errors.
Thanks for your reports!

**Reply by kexul** [reply]
Hi @ptosco , Thanks for your continued helps! Unfortunately, I still got bad case: 
```
O=S(C1=CC(Br)=CC(Cl)=C1O)(NC2=CN(C3CC3)C=N2)=O
O=S(C1=CC(Br)=CC(Cl)=C1O)(NC2=CN(C3CCCCC3)C=N2)=O
```
which raised an error: 
```
rdkit.Chem.rdchem.MolSanitizeException: Explicit valence for atom # 16 H, 2, is greater than permitted
```

**Reply by kexul** [reply]
Hi @ptosco , I've add another condition, it seems ok now: 
```
if b.GetBeginAtom().GetAtomicNum() == 0 and b.GetEndAtom().GetAtomicNum() == 0:
    common_core = Chem.RWMol(common_core)
    common_core.RemoveBond(b.GetBeginAtom().GetIdx(), b.GetEndAtom().GetIdx())
elif b.GetBeginAtom().GetAtomicNum() == 0 or b.GetEndAtom().GetAtomicNum() == 0:
    b.SetBondType(Chem.BondType.SINGLE)
    b.SetIsAromatic(False)
```

**Reply by ptosco** [reply]
Hi @kexul many thanks for reporting this additional issue - great that you found how to fix it yourself.
 I will make sure to update my PR and remove any pre-existing bonds that should not stay in the final molecule. Thanks again for reporting, this is really helpful.

**Reply by kexul** [reply]
@ptosco  You're welcome! Your answer helped me a lot too!
